File: scripts/slop_filter.py
#!/usr/bin/env python3
"""
AI Slop Filter & Text Normalizer (Python)
OpenSchool Blueprint Engine

Normalizes document language and strips out AI clichés, conversational filler, 
and formatting artifacts using compiled regular expressions.
"""

import re
import sys
import json
import argparse
from typing import List, Dict, Any, Tuple

# AI Slop Pattern Catalog
SLOP_PATTERNS = [
    # Conversational Preambles & Meta-Chat
    (re.compile(r'\b(as an ai (language model|assistant)|as a helpful assistant)[^,\.\n]*[,\.]?', re.IGNORECASE), ''),
    (re.compile(r'\b(here is (a|the) (comprehensive|detailed|breakdown|summary|overview)[^:\.\n]*:?)', re.IGNORECASE), ''),
    (re.compile(r'\b(certainly!|sure thing!|of course!|let\'s delve in!|hope this helps!)\b', re.IGNORECASE), ''),
    (re.compile(r'\b(in this section,? we will (explore|examine|delve into|look at))\b', re.IGNORECASE), ''),

    # AI Clichés & Buzzword Fillers
    (re.compile(r'\b(delve(?:s|d)? into(?: the realm of)?)\b', re.IGNORECASE), 'examine'),
    (re.compile(r'\b(a testament to(?: the)?)\b', re.IGNORECASE), 'evidence of'),
    (re.compile(r'\b(in the (?:rapidly )?ever-evolving (?:landscape|world|paradigm) of)\b', re.IGNORECASE), 'in'),
    (re.compile(r'\b(in today\'s fast-paced world,?)\b', re.IGNORECASE), 'currently,'),
    (re.compile(r'\b(tapestry of)\b', re.IGNORECASE), 'range of'),
    (re.compile(r'\b(unlock(?:ing|s|ed)? (?:the )?potential(?: of)?)\b', re.IGNORECASE), 'enabling'),
    (re.compile(r'\b(harness(?:ing|es|ed)? (?:the )?power of)\b', re.IGNORECASE), 'utilizing'),
    (re.compile(r'\b(plays? a (?:crucial|pivotal|vital|key) role in (?:fostering|shaping|driving))\b', re.IGNORECASE), 'supports'),
    (re.compile(r'\b(unwavering commitment(?: to)?)\b', re.IGNORECASE), 'dedication to'),
    (re.compile(r'\b(beacon of (?:excellence|hope|learning))\b', re.IGNORECASE), 'center for learning'),
    (re.compile(r'\b(seamless(?:ly)? integrat(?:ed|ing|ion|es))\b', re.IGNORECASE), 'integrated'),
    (re.compile(r'\b(synerg(?:y|istic|ies))\b', re.IGNORECASE), 'coordination'),
    (re.compile(r'\b(game-changer|paradigm shift)\b', re.IGNORECASE), 'significant advancement'),
    (re.compile(r'\b(holistic (?:approach|methodology|framework))\b', re.IGNORECASE), 'comprehensive approach'),

    # Empty Transitions & Pompous Connectors
    (re.compile(r'\b(it is worth noting that|it\'s worth noting that|it is important to remember that)\b', re.IGNORECASE), 'note that'),
    (re.compile(r'\b(at its core,?)\b', re.IGNORECASE), 'primarily,'),
    (re.compile(r'\b(furthermore,? moreover,?)\b', re.IGNORECASE), 'furthermore,'),
    (re.compile(r'\b(all in all,? in a nutshell,?)\b', re.IGNORECASE), 'in summary,'),
    (re.compile(r'\b(serves as a cornerstone(?: of)?)\b', re.IGNORECASE), 'serves as a foundation of'),
    (re.compile(r'\b(foster(?:s|ing|ed)? a culture of)\b', re.IGNORECASE), 'promote'),
    (re.compile(r'\b(poised to (?:revolutionize|transform))\b', re.IGNORECASE), 'designed to improve')
]

FORMATTING_PATTERNS = [
    (re.compile(r'[\u00A0\u200B\u200C\u200D\u200E\u200F\uFEFF]'), ' '),
    (re.compile(r'\*{3,}'), '**'),
    (re.compile(r'[ \t]+$', re.MULTILINE), ''),
    (re.compile(r'[^\S\r\n]{2,}'), ' '),
    (re.compile(r'\n{3,}'), '\n\n')
]

def clean_text(text: str) -> str:
    """Normalizes formatting and strips AI slop from text string."""
    if not text or not isinstance(text, str):
        return ""

    cleaned = text
    for pattern, repl in FORMATTING_PATTERNS:
        cleaned = pattern.sub(repl, cleaned)

    for pattern, repl in SLOP_PATTERNS:
        def replace_with_case(match):
            m = match.group(0)
            if m and m[0].isupper() and len(repl) > 0:
                return repl[0].upper() + repl[1:]
            return repl
        cleaned = pattern.sub(replace_with_case, cleaned)

    # Post cleanup punctuation and spacing
    cleaned = re.sub(r'\s+([.,!?;:])', r'\1', cleaned)
    cleaned = re.sub(r'([.,])\1+', r'\1', cleaned)
    cleaned = re.sub(r'\(\s+', '(', cleaned)
    cleaned = re.sub(r'\s+\)', ')', cleaned)
    cleaned = re.sub(r'[^\S\r\n]{2,}', ' ', cleaned)
    return cleaned.strip()

def detect_slop(text: str) -> List[Dict[str, Any]]:
    """Detects slop phrases and returns their positions."""
    if not text:
        return []
    matches = []
    for pattern, repl in SLOP_PATTERNS:
        for m in pattern.finditer(text):
            matches.append({
                "phrase": m.group(0),
                "start": m.start(),
                "end": m.end(),
                "suggestion": repl or "[Remove]"
            })
    return sorted(matches, key=lambda x: x["start"])

def clean_json_data(data: Any) -> Tuple[Any, int]:
    """Recursively cleans all string values in a JSON structure."""
    cleaned_count = 0
    if isinstance(data, dict):
        new_dict = {}
        for k, v in data.items():
            cleaned_v, count = clean_json_data(v)
            new_dict[k] = cleaned_v
            cleaned_count += count
        return new_dict, cleaned_count
    elif isinstance(data, list):
        new_list = []
        for item in data:
            cleaned_item, count = clean_json_data(item)
            new_list.append(cleaned_item)
            cleaned_count += count
        return new_list, cleaned_count
    elif isinstance(data, str):
        matches = detect_slop(data)
        if matches:
            return clean_text(data), len(matches)
        return data, 0
    return data, 0

def main():
    parser = argparse.ArgumentParser(description="Clean AI slop and normalize text in Markdown, Text, or JSON files.")
    parser.add_argument("input_file", help="Path to input file")
    parser.add_argument("-o", "--output", help="Path to output file (defaults to stdout)")
    parser.add_argument("--json", action="store_true", help="Process input as JSON data structure")
    args = parser.parse_args()

    try:
        with open(args.input_file, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception as e:
        print(f"Error reading file {args.input_file}: {e}", file=sys.stderr)
        sys.exit(1)

    if args.json or args.input_file.endswith('.json'):
        try:
            data = json.loads(content)
            cleaned_data, count = clean_json_data(data)
            output_str = json.dumps(cleaned_data, indent=2, ensure_ascii=False)
            print(f"Cleaned {count} AI slop occurrences in JSON.", file=sys.stderr)
        except Exception as e:
            print(f"Error parsing JSON: {e}", file=sys.stderr)
            sys.exit(1)
    else:
        matches = detect_slop(content)
        cleaned_content = clean_text(content)
        print(f"Cleaned {len(matches)} AI slop occurrences in text.", file=sys.stderr)
        output_str = cleaned_content

    if args.output:
        with open(args.output, 'w', encoding='utf-8') as f:
            f.write(output_str)
        print(f"Saved cleaned output to {args.output}", file=sys.stderr)
    else:
        print(output_str)

if __name__ == "__main__":
    main()

File: src/engine/slop-filter.js
/**
 * AI Slop Filter & Language Normalization Engine
 * 
 * Cleans formulaic AI filler, clichés, conversational artifacts, 
 * and formatting debris from architectural and educational text descriptions.
 */

// Common AI buzzwords, hyperbolic transitions, and empty filler phrases
export const SLOP_PATTERNS = [
  // Conversational LLM Preambles & Meta-Chat
  { pattern: /\b(as an ai (language model|assistant)|as a helpful assistant)[^,\.\n]*[,\.]?/gi, replacement: '' },
  { pattern: /\b(here is (a|the) (comprehensive|detailed|breakdown|summary|overview)[^:\.\n]*:?)/gi, replacement: '' },
  { pattern: /\b(certainly!|sure thing!|of course!|let's delve in!|hope this helps!)\b/gi, replacement: '' },
  { pattern: /\b(in this section,? we will (explore|examine|delve into|look at))\b/gi, replacement: '' },

  // Overused AI Clichés & Buzzword Fillers
  { pattern: /\b(delve(?:s|d)? into(?: the realm of)?)\b/gi, replacement: 'examine' },
  { pattern: /\b(a testament to(?: the)?)\b/gi, replacement: 'evidence of' },
  { pattern: /\b(in the (?:rapidly )?ever-evolving (?:landscape|world|paradigm) of)\b/gi, replacement: 'in' },
  { pattern: /\b(in today's fast-paced world,?)\b/gi, replacement: 'currently,' },
  { pattern: /\b(tapestry of)\b/gi, replacement: 'range of' },
  { pattern: /\b(unlock(?:ing|s|ed)? (?:the )?potential(?: of)?)\b/gi, replacement: 'enabling' },
  { pattern: /\b(harness(?:ing|es|ed)? (?:the )?power of)\b/gi, replacement: 'utilizing' },
  { pattern: /\b(plays? a (?:crucial|pivotal|vital|key) role in (?:fostering|shaping|driving))\b/gi, replacement: 'supports' },
  { pattern: /\b(unwavering commitment(?: to)?)\b/gi, replacement: 'dedication to' },
  { pattern: /\b(beacon of (?:excellence|hope|learning))\b/gi, replacement: 'center for learning' },
  { pattern: /\b(seamless(?:ly)? integrat(?:ed|ing|ion|es))\b/gi, replacement: 'integrated' },
  { pattern: /\b(synerg(?:y|istic|ies))\b/gi, replacement: 'coordination' },
  { pattern: /\b(game-changer|paradigm shift)\b/gi, replacement: 'significant advancement' },
  { pattern: /\b(holistic (?:approach|methodology|framework))\b/gi, replacement: 'comprehensive approach' },
  { pattern: /\b(rich tapestry|vibrant ecosystem)\b/gi, replacement: 'environment' },

  // Empty Transitionals and Pompous Connectors
  { pattern: /\b(it is worth noting that|it's worth noting that|it is important to remember that)\b/gi, replacement: 'note that' },
  { pattern: /\b(at its core,?)\b/gi, replacement: 'primarily,' },
  { pattern: /\b(furthermore,? moreover,?)\b/gi, replacement: 'furthermore,' },
  { pattern: /\b(all in all,? in a nutshell,?)\b/gi, replacement: 'in summary,' },
  { pattern: /\b(serves as a cornerstone(?: of)?)\b/gi, replacement: 'serves as a foundation of' },
  { pattern: /\b(foster(?:s|ing|ed)? a culture of)\b/gi, replacement: 'promote' },
  { pattern: /\b(poised to (?:revolutionize|transform))\b/gi, replacement: 'designed to improve' }
];

// Formatting and invisible artifact regular expressions
export const FORMATTING_PATTERNS = [
  // Invisible / non-standard Unicode characters
  { pattern: /[\u00A0\u200B\u200C\u200D\u200E\u200F\uFEFF]/g, replacement: ' ' },
  // Excessive Markdown bolding/asterisks (3 or more)
  { pattern: /\*{3,}/g, replacement: '**' },
  // Trailing whitespace on lines
  { pattern: /[ \t]+$/gm, replacement: '' },
  // Multiple consecutive whitespace characters (excluding newlines)
  { pattern: /[^\S\r\n]{2,}/g, replacement: ' ' },
  // Three or more consecutive line breaks to two
  { pattern: /\n{3,}/g, replacement: '\n\n' }
];

/**
 * Detects all slop phrases in a given text string.
 * @param {string} text - The input text to inspect
 * @returns {Array<{phrase: string, index: number, suggestion: string}>}
 */
export function detectSlopMatches(text) {
  if (!text || typeof text !== 'string') return [];
  
  const matches = [];
  
  for (const { pattern, replacement } of SLOP_PATTERNS) {
    const regex = new RegExp(pattern.source, pattern.flags);
    let match;
    while ((match = regex.exec(text)) !== null) {
      matches.push({
        phrase: match[0],
        index: match.index,
        suggestion: replacement.trim() || '[Remove]'
      });
    }
  }

  return matches.sort((a, b) => a.index - b.index);
}

/**
 * Calculates the percentage of "slop density" in a text snippet.
 * @param {string} text 
 * @returns {number} Score from 0 to 100
 */
export function calculateSlopScore(text) {
  if (!text || typeof text !== 'string') return 0;
  const words = text.trim().split(/\s+/).filter(Boolean);
  if (words.length === 0) return 0;

  const matches = detectSlopMatches(text);
  if (matches.length === 0) return 0;

  let slopWordCount = 0;
  matches.forEach(m => {
    slopWordCount += m.phrase.trim().split(/\s+/).length;
  });

  const rawScore = (slopWordCount / words.length) * 100;
  return Math.min(100, Math.round(rawScore * 10) / 10);
}

/**
 * Cleans a text string by removing AI slop and normalizing formatting.
 * @param {string} text - Raw input text
 * @returns {string} Sanitized and normalized text
 */
export function cleanText(text) {
  if (!text || typeof text !== 'string') return text || '';

  let cleaned = text;

  // 1. Run formatting normalizations
  for (const { pattern, replacement } of FORMATTING_PATTERNS) {
    cleaned = cleaned.replace(pattern, replacement);
  }

  // 2. Replace slop phrases
  for (const { pattern, replacement } of SLOP_PATTERNS) {
    cleaned = cleaned.replace(pattern, (match) => {
      // Preserve uppercase if original started with capital
      if (/^[A-Z]/.test(match) && replacement.length > 0) {
        return replacement.charAt(0).toUpperCase() + replacement.slice(1);
      }
      return replacement;
    });
  }

  // 3. Clean up residual double punctuation or broken spacing
  cleaned = cleaned
    .replace(/\s+([.,!?;:])/g, '$1') // Space before punctuation
    .replace(/([.,])\1+/g, '$1')      // Double periods or commas
    .replace(/\(\s+/g, '(')           // Space after open parenthesis
    .replace(/\s+\)/g, ')')           // Space before close parenthesis
    .replace(/[^\S\r\n]{2,}/g, ' ')   // Extra spaces
    .trim();

  return cleaned;
}

/**
 * Recursively scans and cleans an OpenSchool project object.
 * Sanitizes project notes, room notes, staffing descriptions, etc.
 * @param {object} projectData - The full OpenSchool project data model
 * @returns {{ sanitizedProject: object, stats: { totalCleanedFields: number, matchesFound: number } }}
 */
export function sanitizeSchoolProject(projectData) {
  if (!projectData || typeof projectData !== 'object') {
    return { sanitizedProject: projectData, stats: { totalCleanedFields: 0, matchesFound: 0 } };
  }

  let totalCleanedFields = 0;
  let matchesFound = 0;

  const deepClean = (obj) => {
    if (Array.isArray(obj)) {
      return obj.map(item => deepClean(item));
    } else if (obj !== null && typeof obj === 'object') {
      const newObj = {};
      for (const [key, value] of Object.entries(obj)) {
        if (typeof value === 'string') {
          const detected = detectSlopMatches(value);
          if (detected.length > 0) {
            matchesFound += detected.length;
            totalCleanedFields += 1;
            newObj[key] = cleanText(value);
          } else {
            newObj[key] = value;
          }
        } else {
          newObj[key] = deepClean(value);
        }
      }
      return newObj;
    }
    return obj;
  };

  const sanitizedProject = deepClean(projectData);
  return {
    sanitizedProject,
    stats: {
      totalCleanedFields,
      matchesFound
    }
  };
}
